File: c/insertion_sort.c
#include<stdio.h>
void insertion_sort(int* a, int n){
	int i,j;
	for (j=1;j<n;j++){
		int key= a[j];
		int i=j-1;
		while (i>=0 && a[i]>key){
			a[i+1]=a[i];
			i--;
		}
		a[i+1]=key;
	}
}
void printarray(int* a,int n){
	int i;
	printf("unsorted array: ");
	for(i=0;i<n;i++){
		printf("%d ",a[i]);
	}
	printf("\n");
}
int main() {
	int arr[] = {15,9,85,3,4,1,2,10};
	int len = sizeof(arr) / sizeof(arr[0]);
	int i;
	printf("unsorted array: ");
	printarray(arr,len);
	insertion_sort(arr, len);
	printf("Sorted array: ");
	printarray(arr,len);
	return 0;
}

File: java learn/dsa/InsertionSort.java
class InsertionSort{
	public void Insertion(int[] arr){
		for(int i=1;i<arr.length;i++){
			int key=arr[i];
			int j=i-1;
			while(j>=0 && arr[j]>key){
				arr[j+1]=arr[j];
				j=j-1;
			}
			arr[j+1]=key;
		}
	}
	public void recIns(int[] arr,int len){
		if(len<=1) return;
		int key=arr[len-1];
		int j=len-2;
		recIns(arr,len-1);
		while(j>=0 && arr[j]>key){
			arr[j+1]=arr[j];
			j--;
		}
		arr[j+1]=key;
	}



	public static void main(String[] args){
		int[] a={7,3,5,4,2,6};
		int l=a.length;
		InsertionSort obj= new InsertionSort();
		//obj.Insertion(a);
		obj.recIns(a,l);
		for(int x: a){
			System.out.print(x+"  ");
		}
	}
}
